### ISSUE TITLE
android application crashes when detaching the session

### ISSUE THREAD

**Issue by muhzii** [frida/frida#545]
So I can attach to a process and load my script to start hooking a function on the device. But when I try to detach the session or unload the script the application often crashes. I'd want to attach to a process , do some work then detach the session and continue the work-flow of the program.

```
session = frida.get_device_manager().enumerate_devices()[-1].attach(package_name)
script = session.create_script(get_hook_code())
script.on('message', on_message)
script.load()
# do something
script.unload()
session.detach()
# continue program work-flow
```

**Comment by muhzii** [comment]
After doing some steps, I found that crashing only happens if the function is being hooked at the same time the `detach` or `unload` methods are called e.g. the hooked function is being called in a loop inside the application. Is that normal?

**Comment by oleavr** [comment]
What kind of hooking are you referring to?

**Comment by muhzii** [comment]
@oleavr the Java.perform() one

here's an example:
```
setTimeout(function(){
            Java.perform(function () {
                var TM = Java.use("android.telephony.TelephonyManager");
                TM.getDeviceId.overload().implementation = function () {
                    send("deviceID() is called");
                    return "";
                };
            });
        },0);
        """
```

**Comment by oleavr** [comment]
Right. That's a known issue – frida-java is not yet trying to do this in a concurrency-safe manner. PRs welcome!

PS: Remove that `setTimeout()` call, otherwise when using this script with early instrumentation you allow the app to run its own code before your instrumentation is applied, which is probably not what you want.
